mkBinaryCache: improve performance by processing items in parallel

thomasjm opened the pull request:
`mkBinaryCache` can become very slow while building a non-trivial cache, because it processes each item sequentially and because the `xz` compression it uses is somewhat slow.

This PR solves the first problem by using Python's `multiprocessing` library to do the processing in parallel. It gets the number of workers from the `NIX_BUILD_CORES` environment variable. 

This results in a substantial performance improvement: in my testing, the expression `mkBinaryCache { rootPaths = [hello curl emacs htop]; }` goes from **117.4s** to **22.6s** (a **5.2x** speedup).

I also verified that it works on macOS, where using `multiprocessing` can be a little trickier.

The tests pass for me if you rebase this on https://github.com/NixOS/nixpkgs/pull/370773; it might be good to land that one first.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

thomasjm committed:
mkBinaryCache: process items in parallel

thomasjm commented:
Hi @roberth -- you reviewed the original PR for mkBinaryCache, any chance you could take a look at this improvement?

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
Mind adding a commit that runs `ruff format` on this file?

thomasjm replied:
Done

thomasjm commented:
Hi there @philiptaron -- since you were kind of enough to look at my other PR, and I've had trouble getting reviewers for `mkBinaryCache`, is there any way I could trouble you to look at this one as well? It should be a perf improvement only with no change to functionality.

This is the 2nd of 3 PRs I'd like to make to get `mkBinaryCache` to be fast and well-tested.

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,32 +1,30 @@
-  fileHash = subprocess.run(["nix-hash", "--base32", "--type", "sha256", "--flat", xzFile], capture_output=True).stdout.decode().strip()
+  fileHash = subprocess.run(
+    ["nix-hash", "--base32", "--type", "sha256", "--flat", xzFile],
+    capture_output=True
```
needs `check=True` also I think. See also the other `subprocess.run` calls.

thomasjm replied:
Done

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
       "NarHash: " + item["narHash"],
       "NarSize: " + str(item["narSize"]),
-      "References: " + " ".join(dropPrefix(ref) for ref in item["references"]),
+      "References: " + " ".join(dropPrefix(ref, nixPrefix) for ref in item["references"]),
```
All of these strings could be clearer with the f-string syntax.

thomasjm replied:
Done

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
+  parser.add_argument("--num-workers", type=int, default=4, help="Number of worker processes")
+  args = parser.parse_args()
+
+  with open(os.environ["NIX_ATTRS_JSON_FILE"], "r") as f:
+    closures = json.load(f)["closure"]
```
I might put this as a command-line parameter with a default of `os.environ.get("NIX_ATTRS_JSON_FILE")`.

thomasjm replied:
After some reflection I decided to go the opposite direction here. I think piping these variables through in multiple places increases complexity/LOC without adding clarity. This Python script is very much not a reusable tool, it is meant to run in a specific context, a derivation where these environment variables must exist.

I had initially passed in `--num-workers` as an argument, but that was when I was considering making this explicitly configurable as an argument to the derivation. In my latest commit, we read that from the `NIX_BUILD_CORES` environment variable directly. (This also avoids the hard-to-read bash-ism you commented on.)

philiptaron replied:
I think that's probably a good call given the script's rationale.

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
+  with open(os.environ["NIX_ATTRS_JSON_FILE"], "r") as f:
+    closures = json.load(f)["closure"]
+
+  outDir = os.environ["out"]
+  nixPrefix = os.environ["NIX_STORE"]
```
Same with these.

thomasjm replied:
See above

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
+  os.chdir(outDir)
+  os.makedirs("nar", exist_ok=True)
+
+  with open("nix-cache-info", "w") as f:
```
This implicitly uses the current working directory. It looks like that's thrown away in the derivation which makes use of this file. How's it being made use of?

thomasjm replied:
See the `os.chdir` above. It's doing this in the `$out` dir.l

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
+
+  with open("nix-cache-info", "w") as f:
+    f.write("StoreDir: " + nixPrefix + "\n")
+
+  with Pool(processes=args.num_workers) as pool:
+    worker = partial(processItem, nixPrefix=nixPrefix, outDir=outDir)
+    pool.map(worker, closures)
```
I pretty strongly favor a locally defined function over `functools.partial`, since that maintains typing information. We also need to maintain some amount of exception handling information.

Here's a sketch of what I'm thinking of, taking some care to maintain the Python stack trace (which actually might not be needed in this case, since process_item is so simple.)

```suggestion

  @dataclass(frozen=True)
  class ProcessError:
     exception: Exception
     traceback: str
     
     def __str__(self):
        return f"{self.e.__class__.__name__}: {self.e}\n{self.traceback}"
     
  def perform_work_in_process(item):
     try:
        processItem(item, nixPrefix=nixPrefix, outDir=outDir)
        return None
     except Exception as e:
        return ProcessError(e, traceback.format_exc())

  with Pool(processes=args.num_workers) as pool:
    async_results = [pool.apply_async(lambda: perform_work_in_process(item=item) for item in closures]
    results = [r.get() for r in results]
    errors = [e for e in results if e is not None]
    pool.join()
    if len(errors) > 0:
       print("Encountered these errors:")
       for error in errors:
          print(str(error))
          
       return 1
    else:
      return 0
```

thomasjm replied:
This proposal turns 3 lines into almost 30 for an advantage I don't really see. As you say, the script is very simple--it pretty much just hashes lots of files and writes the results. If there is an exception, I already know what the stack trace will be: `main -> processItem`. I think it's very easy to debug as-is and I would like to keep it simple.

philiptaron replied:
Yeah, I hear that. Let me resolve this and make a more simple suggestion.

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -39,5 +37,29 @@ def dropPrefix(path):
+    pool.map(worker, closures)
+
+if __name__ == "__main__":
+  main()
```
```suggestion
  sys.exit(main())
```

thomasjm replied:
Hmm, I haven't seen this pattern before. I think this actually just makes more confusing since `main` here doesn't return a value. Python will already exit with a nonzero exit code if there's an unhandled exception.

thomasjm replied:
Oh I see, you were returning an exit code in your sketch for the Pool invocation

philiptaron replied:
Yeah, that's right. I hate Python tracebacks when some subsidiary tool failed, so I try to be well behaved when there are known errors.

philiptaron reviewed (changes requested):
Requesting changes for the error handling in the subprocess calls and in the multiprocessing pool. This tool should not error silently!

philiptaron commented:
> Hi there @philiptaron -- since you were kind enough to look at my other PR, and I've had trouble getting reviewers for `mkBinaryCache`, is there any way I could trouble you to look at this one as well?

On it. I've had a lot of experience with multiprocessing pools so you got the "watch out for dragons" review. 😬

thomasjm commented on pkgs/build-support/binary-cache/default.nix:
```diff
@@ -39,7 +39,10 @@ stdenv.mkDerivation {
     mkdir -p $out/nar
 
-    python ${./make-binary-cache.py}
+    NUM_WORKERS="''${NIX_BUILD_CORES:-4}"
```
Right

thomasjm commented on pkgs/build-support/binary-cache/default.nix:
```diff
@@ -39,7 +39,10 @@ stdenv.mkDerivation {
-    python ${./make-binary-cache.py}
+    NUM_WORKERS="''${NIX_BUILD_CORES:-4}"
+
+    python ${./make-binary-cache.py} \
+      --num-workers "$NUM_WORKERS"
```
Removed this arg

thomasjm commented:
Thanks very much for the review @philiptaron!

I had missed the possibility of silent erroring, so adding `check=True` was a great catch.

A few of the comments I pushed back against in the name of simplicity/conciseness. Let me know what you think!

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+    xzFile = "nar/" + narInfoHash + ".nar.xz"
+    with open(xzFile, "w") as f:
+        subprocess.run(
+            "nix-store --dump %s | xz -c" % item["path"],
```
```suggestion
            f"nix-store --dump {item['path']} | xz -c",
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+    narInfoHash = dropPrefix(item["path"], nixPrefix).split("-")[0]
+
+    xzFile = "nar/" + narInfoHash + ".nar.xz"
+    with open(xzFile, "w") as f:
```
I think this wants binary mode

```suggestion
    with open(xzFile, "wb") as f:
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+    )
+    fileSize = os.path.getsize(xzFile)
+
+    finalXzFile = "nar/" + fileHash + ".nar.xz"
```
```suggestion
    finalXzFile = f"nar/{fileHash}.nar.xz"
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+    finalXzFile = "nar/" + fileHash + ".nar.xz"
+    os.rename(xzFile, finalXzFile)
+
+    with open(narInfoHash + ".narinfo", "w") as f:
```
Text-mode is the default, but this shows the code means to do so.

```suggestion
    with open(f"{narInfoHash}.narinfo", "wt") as f:
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+    os.rename(xzFile, finalXzFile)
+
+    with open(narInfoHash + ".narinfo", "w") as f:
+        f.writelines(
+            (
+                x + "\n"
+                for x in [
+                    f"StorePath: {item['path']}",
+                    f"URL: {finalXzFile}",
+                    "Compression: xz",
+                    f"FileHash: sha256:{fileHash}",
+                    f"FileSize: {fileSize}",
+                    f"NarHash: {item['narHash']}",
+                    f"NarSize: {item['narSize']}",
+                    f"References: {' '.join(dropPrefix(ref, nixPrefix) for ref in item['references'])}",
+                ]
+            )
+        )
```
A little [code golf](https://en.wikipedia.org/wiki/Code_golf).
```suggestion
        f.write(f"StorePath: {item['path']}\n")
        f.write(f"URL: {finalXzFile}\n")
        f.write("Compression: xz\n")
        f.write(f"FileHash: sha256:{fileHash}\n")
        f.write(f"FileSize: {fileSize}\n")
        f.write(f"NarHash: {item['narHash']}\n")
        f.write(f"NarSize: {item['narSize']}\n")
        f.write(f"References: {' '.join(dropPrefix(ref, nixPrefix) for ref in item['references'])}\n")
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
-  with open(xzFile, "w") as f:
-    subprocess.run("nix-store --dump %s | xz -c" % item["path"], stdout=f, shell=True)
+    with open("nix-cache-info", "w") as f:
+        f.write("StoreDir: " + nixPrefix + "\n")
```
```suggestion
        f.write(f"StoreDir: {nixPrefix}\n")
```

philiptaron commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
-  fileSize = os.path.getsize(xzFile)
+    with Pool(processes=numWorkers) as pool:
+        worker = partial(processItem, nixPrefix=nixPrefix, outDir=outDir)
+        pool.map(worker, closures)
```
If we did the strings-for-errors, it would look like this.

```suggestion
        results = pool.map(worker, closures)
        for result in results:
           if result is not None:
               print(result, file=sys.stderr, flush=True)
        return 0 if all(r is None for r in results) else 1
```

philiptaron reviewed (approved):
No blocking comments, just some doodling around the edges. The error case I keep thinking about is these Nix calls failing, but if they do so in practice, I'll submit the PR to move them to a string-based error reporting myself, I think.

philiptaron commented:
I'm planning on merging this after giving @thomasjm a chance to respond.

thomasjm commented on pkgs/build-support/binary-cache/make-binary-cache.py:
```diff
@@ -1,43 +1,78 @@
+
+
+def processItem(item, nixPrefix, outDir):
+    os.chdir(outDir)
```
> Since you're passing in the same path each time, and the process has changed local directory before entering the pool, it's also not necessary.

I think that may not be true on macOS, which uses separate processes for its threadpools. I'd have to test to be sure.

But, what we can do is just not depend on the pwd. Let me update this to explicitly join the `outDir` onto all paths we deal with.

thomasjm replied:
Done

thomasjm commented:
Thanks again @philiptaron! I accepted most of the comments.

Removing all `os.chdirs` was a nice change, you can see how I did that by explicitly constructing all paths.

Still not really feeling the error reporting stuff. As a test I introduced an error into `processItem` and ran it, and I got a perfectly comprehensible stack trace. `Pool.map` will rethrow exceptions from workers after all, I don't see why we need more than that.

I just tried out the latest version on my macOS machine and it still works.

Just squashed everything, I think it's ready to go.

philiptaron commented:
Tested it with `nix-build -A nixosTests.binary-cache`.
